knative-operator/pkg/common/service_monitor.go

```diff
@@ -54,23 +81,210 @@ func SetupServiceMonitor(ctx context.Context, cfg *rest.Config, metricsPort int3
 // serveCRMetrics gets the Operator/CustomResource GVKs and generates metrics based on those types.
 // It serves those metrics on "http://metricsHost:operatorMetricsPort".
 func serveCRMetrics(cfg *rest.Config, metricsHost string, operatorMetricsPort int32) error {
-	// Below function returns filtered operator/CustomResource specific GVKs.
-	// For more control override the below GVK list with your own custom logic.
-	filteredGVK, err := k8sutil.GetGVKsFromAddToScheme(apis.AddToScheme)
+
+	// If we dont use a custom list here, the typical call to get a filtered list of gvks using k8sutil.GetGVKsFromAddToScheme(apis.AddToScheme)
+	// will end up with resources the Operator does not have access to eg. `Kind=LimitRange`
+	// The list of resources returned is unrelated to our purposes here, thus the customization.
+	gvkFilterList := []schema.GroupVersionKind{
+		schema.GroupVersionKind{
+			Group:   "operator.knative.dev",
+			Version: "v1alpha1",
+			Kind:    "KnativeServing",
+		},
+		schema.GroupVersionKind{
+			Group:   "operator.knative.dev",
+			Version: "v1alpha1",
+			Kind:    "KnativeEventing",
+		},
+	}
+	// To generate metrics in other namespaces, add the values below.
+	// This is due to this bug: https://github.com/operator-framework/operator-sdk/issues/2494
+	// For the workaround check here: https://github.com/operator-framework/operator-sdk/pull/2601/files#r396745465
+	// and https://github.com/shipwright-io/build/pull/73
+	// In order to avoid getting a bad value we avoid using k8sutil.GetWatchNamespace() that gets the value from the WATCH_NAMESPACE
+	// env var. That value is by default "" but user may change it, affecting the metrics endpoint.
+	namespaces := []string{""}
+	// Generate and serve custom resource specific metrics.
+	err := kubemetrics.GenerateAndServeCRMetrics(cfg, namespaces, gvkFilterList, metricsHost, operatorMetricsPort)
 	if err != nil {
 		return err
 	}
-	// Get the namespace the operator is currently deployed in.
-	operatorNs, err := k8sutil.GetOperatorNamespace()
+	return nil
+}
+
+func SetupEventingBrokerServiceMonitors(client client.Client, namespace string, instance *eventingv1alpha1.KnativeEventing) error {
+	if err := isServiceMonitorAvailable(); err != nil {
```

Comment:
Is there a possibility that this is ever false in an Openshift cluster? If not, I'd drop this.

Reply:
I guess w/ ocp it's always there, yes

Reply:
This is what the coreos monitoring code does so I thought it would be safer to do keep check [here](https://github.com/operator-framework/operator-sdk/blob/0258db0119e8e18e15d035532427c329fce1e871/pkg/metrics/service-monitor.go#L114).

Reply:
Anyway I will drop it.